# Preparation

Preparation biasa dilakukan untuk mempersiapkan data sebelum masuk dalam tahap pemodelan. <br>
Berikut adalah tahapan yang akan dilalui pada data `SC_HW1_bank_data.csv` sebelum tahap pemodelan :
1. Import Library
2. Input Dataset
3. Preprocessing
4. Train-Test Split

## Import Library

In [ ]:
import pandas as pd
import numpy as np

## Input Dataset

In [ ]:
#Membaca data dan memasukkannya ke dalam bentuk Pandas Dataframe
df = pd.read_csv('https://raw.githubusercontent.com/Rietaros/kampus_merdeka/main/SC_HW1_bank_data.csv')

In [ ]:
#Jalankan code untuk mengecek nama kolom yang tersedia
df.columns

Index(['RowNumber', 'CustomerId', 'Geography', 'Gender', 'Age', 'Tenure',
       'Balance', 'NumOfProducts', 'HasCrCard', 'IsActiveMember',
       'EstimatedSalary', 'Exited'],
      dtype='object')

In [ ]:
df.head()

RowNumber  CustomerId Geography  Gender  Age  Tenure    Balance  \
0          1    15634602    France  Female   42       2       0.00   
1          2    15647311     Spain  Female   41       1   83807.86   
2          3    15619304    France  Female   42       8  159660.80   
3          4    15701354    France  Female   39       1       0.00   
4          5    15737888     Spain  Female   43       2  125510.82   

   NumOfProducts  HasCrCard  IsActiveMember  EstimatedSalary  Exited  
0              1          1               1        101348.88       1  
1              1          0               1        112542.58       0  
2              3          1               0        113931.57       1  
3              2          0               0         93826.63       0  
4              1          1               1         79084.10       0

In [ ]:
#Hilangkan kolom yang dirasa tidak relevan dengan model (contoh: ID). None dapat diisi dengan nama-nama kolom yang akan digunakan.
#Contoh df = df[['X1','X2', 'Y']].copy()

#START CODE
df = df[['Geography', 'Gender', 'Age', 'Tenure', 'Balance', 'NumOfProducts', 'HasCrCard', 'IsActiveMember', 'EstimatedSalary', 'Exited']].copy()
df

Geography  Gender  Age  Tenure    Balance  NumOfProducts  HasCrCard  \
0       France  Female   42       2       0.00              1          1   
1        Spain  Female   41       1   83807.86              1          0   
2       France  Female   42       8  159660.80              3          1   
3       France  Female   39       1       0.00              2          0   
4        Spain  Female   43       2  125510.82              1          1   
...        ...     ...  ...     ...        ...            ...        ...   
9995    France    Male   39       5       0.00              2          1   
9996    France    Male   35      10   57369.61              1          1   
9997    France  Female   36       7       0.00              1          0   
9998   Germany    Male   42       3   75075.31              2          1   
9999    France  Female   28       4  130142.79              1          1   

      IsActiveMember  EstimatedSalary  Exited  
0                  1        101348.88       1  
1                  1        112542.58       0  
2                  0        113931.57       1  
3                  0         93826.63       0  
4                  1         79084.10       0  
...              ...              ...     ...  
9995               0         96270.64       0  
9996               1        101699.77       0  
9997               1         42085.58       1  
9998               0         92888.52       1  
9999               0         38190.78       0  

[10000 rows x 10 columns]

## Preprocessing

In [ ]:
#Lakukan One-Hot Encoder untuk data categorical, dengan fungsi pandas get_dummies

#START CODE
df = pd.get_dummies(df, columns=['Geography', 'Gender'], drop_first=True)

df

Age  Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0      42       2       0.00              1          1               1   
1      41       1   83807.86              1          0               1   
2      42       8  159660.80              3          1               0   
3      39       1       0.00              2          0               0   
4      43       2  125510.82              1          1               1   
...   ...     ...        ...            ...        ...             ...   
9995   39       5       0.00              2          1               0   
9996   35      10   57369.61              1          1               1   
9997   36       7       0.00              1          0               1   
9998   42       3   75075.31              2          1               0   
9999   28       4  130142.79              1          1               0   

      EstimatedSalary  Exited  Geography_Germany  Geography_Spain  Gender_Male  
0           101348.88       1                  0                0            0  
1           112542.58       0                  0                1            0  
2           113931.57       1                  0                0            0  
3            93826.63       0                  0                0            0  
4            79084.10       0                  0                1            0  
...               ...     ...                ...              ...          ...  
9995         96270.64       0                  0                0            1  
9996        101699.77       0                  0                0            1  
9997         42085.58       1                  0                0            0  
9998         92888.52       1                  1                0            1  
9999         38190.78       0                  0                0            0  

[10000 rows x 11 columns]

In [ ]:
#Pisahkan mana X (feature) dengan Y,
#Y adalah kolom "Exited"

#START CODE
X = df.drop(columns=['Exited'])  # Drop the 'Exited' column to get X
y = df['Exited']  # Extract the 'Exited' column as Y

In [ ]:
# Display X
print("Features (X):")
X.head()

Features (X):


Age  Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0   42       2       0.00              1          1               1   
1   41       1   83807.86              1          0               1   
2   42       8  159660.80              3          1               0   
3   39       1       0.00              2          0               0   
4   43       2  125510.82              1          1               1   

   EstimatedSalary  Geography_Germany  Geography_Spain  Gender_Male  
0        101348.88                  0                0            0  
1        112542.58                  0                1            0  
2        113931.57                  0                0            0  
3         93826.63                  0                0            0  
4         79084.10                  0                1            0

In [ ]:
# Display y
print("Target Variable (Y):")
y.head()

Target Variable (Y):


0    1
1    0
2    1
3    0
4    0
Name: Exited, dtype: int64

In [ ]:
#Lakukan Scaler dan/atau Noermalisasi jika diperlukan
from sklearn.preprocessing import MinMaxScaler

# Initialize the Min-Max scaler
scaler = MinMaxScaler()

# Fit and transform the scaler on your feature data (X)
X_transformed = scaler.fit_transform(X)

In [ ]:
#Ini digunakan jika dilakukan scaler/Normalisas. Jika tidak, code ini bisa dilewat dan diganti dengan code yang ada di dalam komen
X_transform = pd.DataFrame(X_transformed, columns = X.columns)
#X_transform = X.copy()

## Train-Test Split

In [ ]:
#Split menjadi train dan test dengan test_size 25%
#Tidak perlu mengubah code ini

from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test = train_test_split(X_transformed,y,test_size = 0.25,random_state = 123)

# Modeling

## Model1
### Soal :
Jelaskan secara Singkat Model ke-1 yang digunakan!

**LOGISTIC REGRESSION**

**Logistic Regression digunakan untuk memprediksi probabilitas sukses atau kegagalan suatu peristiwa berdasarkan satu atau beberapa variabel input.
Output dari model Logistic Regression adalah probabilitas dalam rentang 0 hingga 1, yang dapat diinterpretasikan sebagai kemungkinan kejadian tertentu. Dalam klasifikasi biner, model ini membagi data menjadi dua kelas (biasanya "1" dan "0") berdasarkan probabilitas prediksi. Untuk menghasilkan prediksi, Logistic Regression menggunakan fungsi logistik (sigmoid) yang melengkung yang mengubah hasil dari kombinasi linear fitur menjadi probabilitas. Namun, logistic regression memiliki batasan ketika menangani masalah dengan kompleksitas tinggi atau ketika ada hubungan non-linear antara fitur dan target**

In [ ]:
#Pilih salah satu metode Machine Leaarning
#Model Machine Learning dapat dipanggil terlebih dahulu melalui library yang digunakan. Gunakan library scikit learn seperti pada contoh

# Import the machine learning model you want to use, for example, LogisticRegression
from sklearn.linear_model import LogisticRegression

# Initialize the model
model1 = LogisticRegression()

# Define the hyperparameters and their possible values
params = {
    "tol": [0.1, 0.01, 0.001],
    'C': [0.5, 1.0, 1.5, 2.0]
}

# Import necessary libraries for hyperparameter tuning
from sklearn.model_selection import GridSearchCV

# Create a GridSearchCV object
grid = GridSearchCV(
    estimator=model1,
    param_grid=params,
    scoring='accuracy',
    n_jobs=10,  # Number of CPU cores to use
    cv=10  # Number of cross-validation folds
)

# Fit the model on your training data
grid.fit(X_train, y_train)

# Get the best hyperparameters from the grid search
best_params = grid.best_params_
print("Best Hyperparameters:", best_params)


Best Hyperparameters: {'C': 0.5, 'tol': 0.01}


In [ ]:
#lakukan evaluasi dengan beberapa maetric di bawah ini
from sklearn.metrics import classification_report
from sklearn.metrics import accuracy_score
from sklearn.metrics import confusion_matrix

y_pred = grid.predict(X_test)

print(classification_report(y_test,y_pred))
print("")
print(confusion_matrix(y_test,y_pred))
print("")
print(accuracy_score(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.82      0.97      0.89      1983
           1       0.66      0.20      0.31       517

    accuracy                           0.81      2500
   macro avg       0.74      0.59      0.60      2500
weighted avg       0.79      0.81      0.77      2500


[[1929   54]
 [ 413  104]]

0.8132


##Model 2
### Soal :
Jelaskan secara Singkat Model ke-2 yang digunakan!

**DECISION TREE**


**Decision Tree adalah model yang mengambil bentuk pohon, dengan node akar sebagai awalnya dan node-node daun sebagai hasil akhirnya. Hasil pemilihan fitur tersebut mengarahkan model ke node anak yang sesuai, dan proses ini berlanjut hingga mencapai node daun, di mana keputusan atau prediksi akhir dibuat. Decision Tree mudah dipahami dan dapat diinterpretasikan, cocok untuk data dengan fitur kategorikal atau numerik, dan digunakan baik untuk masalah klasifikasi maupun regresi. Namun, mereka dapat cenderung overfitting jika tidak diatur dengan baik dan sensitif terhadap perubahan kecil dalam data pelatihan.**



In [ ]:
from sklearn.tree import DecisionTreeClassifier

# Initialize the Decision Tree model
model2 = DecisionTreeClassifier()

# Define the hyperparameters for the Decision Tree Classifier
params = {
    "criterion": ["gini", "entropy"],
    "max_depth": [None, 10, 20, 30],
    "min_samples_split": [2, 5, 10],
    "min_samples_leaf": [1, 2, 4]
}

# Import library yang diperlukan untuk melakukan GridSearchCV
from sklearn.model_selection import GridSearchCV

# Buat objek GridSearchCV
grid2 = GridSearchCV(
    estimator=model2,
    param_grid=params,
    scoring='accuracy',
    n_jobs=10,  # Jumlah core CPU yang digunakan
    cv=10  # Jumlah lipatan cross-validation
)

# Lakukan penyetelan parameter pada data pelatihan
grid2.fit(X_train, y_train)

# Dapatkan hyperparameter terbaik dari hasil penyetelan
best_params = grid2.best_params_
print("Best Hyperparameters:", best_params)


Best Hyperparameters: {'criterion': 'gini', 'max_depth': 10, 'min_samples_leaf': 4, 'min_samples_split': 2}


In [ ]:
#lakukan evaluasi dengan beberapa maetric di bawah ini
from sklearn.metrics import classification_report
from sklearn.metrics import accuracy_score
from sklearn.metrics import confusion_matrix

y_pred = grid2.predict(X_test)

print(classification_report(y_test,y_pred))
print("")
print(confusion_matrix(y_test,y_pred))
print("")
print(accuracy_score(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.88      0.95      0.91      1983
           1       0.72      0.49      0.58       517

    accuracy                           0.85      2500
   macro avg       0.80      0.72      0.75      2500
weighted avg       0.84      0.85      0.84      2500


[[1886   97]
 [ 266  251]]

0.8548


##Model 3
### Soal :
Jelaskan secara Singkat Model ke-3 yang digunakan!

 **GRADIENTBOOSTING CLASSIFIER**
 **GradientBoostingClassifier adalah bagian dari keluarga algoritma "Boosting," yang berarti menggabungkan beberapa model lemah menjadi satu model kuat. Algoritma ini bekerja dengan cara secara berurutan membangun serangkaian Decision Trees kecil yang disebut "weak learners" dan memperbaiki kesalahan prediksi sebelumnya pada setiap langkah. Gabungan dari serangkaian Decision Trees ini membentuk model yang kuat yang dapat digunakan untuk klasifikasi. GradientBoostingClassifier menggunakan teknik gradien (gradient descent) untuk mengoptimalkan modelnya dengan mengurangi kesalahan prediksi secara berangsur-angsur. Namun, perlu diatur dengan hati-hati untuk menghindari overfitting, dan dapat memerlukan waktu yang lebih lama untuk melatih dibandingkan dengan model lainnya.**

In [ ]:
from sklearn.ensemble import GradientBoostingClassifier

# Initialize the Gradient Boosting Classifier
model = GradientBoostingClassifier(n_estimators=300, max_depth=3, random_state=42)

# Fit the model on the training data
model.fit(X_train, y_train)

# Make predictions on the test data
y_pred = model.predict(X_test)

In [ ]:
#lakukan evaluasi
from sklearn.metrics import classification_report
from sklearn.metrics import accuracy_score
from sklearn.metrics import confusion_matrix

y_pred = model.predict(X_test)

print(classification_report(y_test,y_pred))
print("")
print(confusion_matrix(y_test,y_pred))
print("")
print(accuracy_score(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.88      0.96      0.92      1983
           1       0.76      0.51      0.61       517

    accuracy                           0.87      2500
   macro avg       0.82      0.74      0.77      2500
weighted avg       0.86      0.87      0.86      2500


[[1899   84]
 [ 251  266]]

0.866


##Model 4
### Soal :
Jelaskan secara Singkat Model ke-4 yang digunakan!

**KNEIGHBORS CLASSIFIER**

**KNeighborsClassifier adalah algoritma berbasis instansi (instance-based) yang digunakan untuk masalah klasifikasi. Algoritma ini bekerja dengan cara "mengenal tetangga" dari setiap data pelatihan dan memutuskan kelas dari data uji berdasarkan mayoritas kelas tetangga terdekatnya. Kelemahan KNeighborsClassifier adalah kinerjanya yang lambat jika datasetnya besar, karena harus memeriksa jarak ke setiap titik data pelatihan. Algoritma ini juga sensitif terhadap skala data, sehingga preprocessing seperti penskalaan fitur seringkali diperlukan.**

In [ ]:
from sklearn.neighbors import KNeighborsClassifier

# Initialize the K-Nearest Neighbors (KNN) model
model4 = KNeighborsClassifier()

# Define the hyperparameters for the KNN Classifier
params = {
    "n_neighbors": [5, 15, 20, 5],
    "weights": ["uniform", "distance"],
    "algorithm": ["auto", "ball_tree", "kd_tree", "brute"]
}

# Import library yang diperlukan untuk melakukan GridSearchCV
from sklearn.model_selection import GridSearchCV

# Buat objek GridSearchCV
grid4 = GridSearchCV(
    estimator=model4,
    param_grid=params,
    scoring='accuracy',
    n_jobs=10,  # Jumlah core CPU yang digunakan
    cv=10  # Jumlah lipatan cross-validation
)

# Lakukan penyetelan parameter pada data pelatihan
grid4.fit(X_train, y_train)

# Dapatkan hyperparameter terbaik dari hasil penyetelan
best_params = grid4.best_params_
print("Best Hyperparameters:", best_params)

Best Hyperparameters: {'algorithm': 'auto', 'n_neighbors': 5, 'weights': 'uniform'}


In [ ]:
#lakukan evaluasi dengan beberapa maetric di bawah ini
from sklearn.metrics import classification_report
from sklearn.metrics import accuracy_score
from sklearn.metrics import confusion_matrix

y_pred = grid4.predict(X_test)

print(classification_report(y_test,y_pred))
print("")
print(confusion_matrix(y_test,y_pred))
print("")
print(accuracy_score(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.85      0.95      0.90      1983
           1       0.63      0.36      0.46       517

    accuracy                           0.82      2500
   macro avg       0.74      0.65      0.68      2500
weighted avg       0.80      0.82      0.80      2500


[[1876  107]
 [ 333  184]]

0.824


##Model 5
### Soal :
Jelaskan secara Singkat Model ke-5 yang digunakan!

**RANDOM FOREST CLASSIFIER**

**RandomForestClassifier adalah algoritma ensemble learning yang membangun sejumlah besar Decision Trees (pohon keputusan) dan menggabungkan hasil prediksi dari pohon-pohon tersebut. Setiap pohon dalam Random Forest dibangun dengan menggunakan subset acak dari data pelatihan (bootstrap) dan subset acak dari fitur (subsetting). Algoritma ini efektif dalam mengatasi overfitting karena penggunaan banyak pohon yang beragam. Algoritma ini cocok untuk berbagai jenis data dan masalah klasifikasi, termasuk data dengan fitur kategorikal atau numerik. Meskipun Random Forest kuat, pengguna harus memperhatikan beberapa hiperparameter seperti jumlah pohon dan kedalaman pohon untuk mengoptimalkan performa model.**

In [ ]:
from sklearn.ensemble import RandomForestClassifier
# Initialize the Random Forest Classifier
model = RandomForestClassifier(n_estimators=100, max_depth=20, random_state=42)

# Fit the model on the training data
model.fit(X_train, y_train)

# Make predictions on the test data
y_pred = model.predict(X_test)

In [ ]:
#lakukan evaluasi dengan beberapa maetric di bawah ini
from sklearn.metrics import classification_report
from sklearn.metrics import accuracy_score
from sklearn.metrics import confusion_matrix

y_pred = model.predict(X_test)

print(classification_report(y_test,y_pred))
print("")
print(confusion_matrix(y_test,y_pred))
print("")
print(accuracy_score(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.88      0.96      0.92      1983
           1       0.77      0.48      0.59       517

    accuracy                           0.86      2500
   macro avg       0.82      0.72      0.75      2500
weighted avg       0.85      0.86      0.85      2500


[[1908   75]
 [ 269  248]]

0.8624


## Tarik Kesimpulan Model Mana yang terbaik beserta alasannya

Dari hasil output yang diberikan, mari kita tinjau masing-masing model dan tarik kesimpulan tentang model mana yang terbaik:

1. Model 1 (Logistic Regression):<br>
Akurasi: 0.8132 atau 81%<br>
Precision (kelas 1): 0.66<br>
Recall (kelas 1): 0.20<br>
F1-score (kelas 1): 0.31<br>

2. Model 2 (Decision Tree):<br>
Akurasi: 0.8548 atau 85%<br>
Precision (kelas 1): 0.72<br>
Recall (kelas 1): 0.49<br>
F1-score (kelas 1): 0.58<br>

3. Model 3 (Gradient Boosting Classifier):<br>
Akurasi: 0.866 atau 87%<br>
Precision (kelas 1): 0.76<br>
Recall (kelas 1): 0.51<br>
F1-score (kelas 1): 0.61<br>

4. Model 4 (KNeighbors Classifier):<br>
Akurasi: 0.824 atau 82%<br>
Precision (kelas 1): 0.63<br>
Recall (kelas 1): 0.36<br>
F1-score (kelas 1): 0.45<br>

5. Model 5 (Random Forest Classifier):<br>
Akurasi: 0.8624 atau 86%<br>
Precision (kelas 1): 0.77<br>
Recall (kelas 1): 0.48<br>
F1-score (kelas 1): 0.59<br>

Berdasarkan hasil di atas, dapat kita lihat bahwa Model 3 (Gradient Boosting Classifier) memiliki akurasi yang lebih tinggi dan skor F1 yang lebih baik untuk kelas 1 dibandingkan dengan model lainnya. Model 3 juga memiliki nilai recall yang lebih tinggi untuk kelas 1, yang menunjukkan bahwa mereka lebih baik dalam mengidentifikasi kasus positif.

Dengan demikian, berdasarkan hasil diatas maka **Model 3 (Gradient Boosting Classifier) mungkin dianggap sebagai model terbaik dengan tingkat akurasi 87 %**.